File: [I]stack-queue/[I]stack-queue/232-implement-queue-using-stacks.py
class MyQueue:

    def __init__(self):
        self.stack1 = []
        self.stack2 = []

    def push(self, x: int) -> None:
        self.stack1.append(x)

    def pop(self) -> int:
        self.peek()
        return self.stack2.pop()

    def peek(self) -> int:
        if self.stack2:
            return self.stack2[- 1]
        while self.stack1:
            self.stack2.append(self.stack1.pop())
        return self.stack2[- 1]

    def empty(self) -> bool:
        return len(self.stack1) == 0 and len(self.stack2) == 0

# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

# time O(1) for push(), empty() ,and initiation, amortized O(1) for pop(), peek()
# space O(n), due to two stacks
# using stack and queue and implement stack/queue and two stacks

File: [I]stack-queue/[I]stack-queue/232-implement-queue-using-stacks.java
class MyQueue {
    ArrayDeque<Integer> stack1;
    ArrayDeque<Integer> stack2;

    public MyQueue() {
        stack1 = new ArrayDeque<>();
        stack2 = new ArrayDeque<>();
    }
    
    public void push(int x) {
        stack1.push(x);
    }
    
    public int pop() {
        peek();
        return stack2.pop();
    }
    
    public int peek() {
        if (!stack2.isEmpty()) {
            return stack2.peek();
        }
        while (!stack1.isEmpty()) {
            stack2.push(stack1.pop());
        }
        return stack2.peek();
    }
    
    public boolean empty() {
        return stack1.isEmpty() && stack2.isEmpty();
    }
}

/**
* Your MyQueue object will be instantiated and called as such:
* MyQueue obj = new MyQueue();
* obj.push(x);
* int param_2 = obj.pop();
* int param_3 = obj.peek();
* boolean param_4 = obj.empty();
*/

// time O(1) for push(), empty() ,and initiation, amortized O(1) for pop(), peek()
// space O(n), due to two stacks
// using stack and queue and implement stack/queue and two stacks